File: Cardiovascular Disease Prediction using Machine Learning and Clinical Data/scripts/r/bivariate_analysis.R
# Bivariate Analysis
#
# This script performs bivariate analysis on the Cardiovascular Disease Dataset.

# Load necessary libraries
library(tidyverse)
library(ggplot2)
library(dplyr)
library(corrplot)
library(gridExtra)

# Set theme for plots
theme_set(theme_minimal() + theme(plot.title = element_text(size = 14, face = "bold")))

# Load the dataset
df <- read.csv("../../data/Cardiovascular_Disease_Dataset.csv", stringsAsFactors = FALSE)

# Correlation Analysis
cat(paste(rep("=", 80), collapse = ""), "\n")
cat("CORRELATION ANALYSIS\n")
cat(paste(rep("=", 80), collapse = ""), "\n")

numerical_cols <- c("age", "restingBP", "serumcholestrol", "maxheartrate", "oldpeak")
correlation_matrix <- cor(df[numerical_cols], use = "complete.obs")

cat("\nCorrelation Matrix:\n")
print(round(correlation_matrix, 3))

# Save correlation plot
png("../../results/correlation_matrix.png", width = 800, height = 600)
corrplot(correlation_matrix, method = "color", type = "upper", 
         order = "hclust", tl.cex = 0.8, tl.col = "black",
         addCoef.col = "black", number.cex = 0.7)
dev.off()

# Numerical vs Target
cat("\n\n", paste(rep("=", 80), collapse = ""), "\n")
cat("NUMERICAL VARIABLES VS TARGET\n")
cat(paste(rep("=", 80), collapse = ""), "\n")

for (col in numerical_cols) {
  if (col %in% names(df)) {
    cat("\n", toupper(col), "by Target:\n")
    grouped <- df %>%
      group_by(target) %>%
      summarise(
        mean = mean(.data[[col]], na.rm = TRUE),
        median = median(.data[[col]], na.rm = TRUE),
        sd = sd(.data[[col]], na.rm = TRUE)
      )
    print(grouped)
    
    # Statistical test
    group_0 <- df[df$target == 0, col]
    group_1 <- df[df$target == 1, col]
    test_result <- t.test(group_0, group_1)
    cat("  t-test p-value:", test_result$p.value, 
        ifelse(test_result$p.value < 0.05, "(Significant)", "(Not significant)"), "\n")
  }
}

# Categorical vs Target
cat("\n\n", paste(rep("=", 80), collapse = ""), "\n")
cat("CATEGORICAL VARIABLES VS TARGET\n")
cat(paste(rep("=", 80), collapse = ""), "\n")

categorical_vars <- c("gender", "chestpain", "fastingbloodsugar", "restingrelectro", 
                     "exerciseangia", "slope", "noofmajorvessels")

for (var in categorical_vars) {
  if (var %in% names(df)) {
    cat("\n", toupper(var), "vs TARGET:\n")
    crosstab <- table(df[[var]], df$target)
    print(crosstab)
    
    # Chi-square test
    test_result <- chisq.test(crosstab)
    cat("  Chi-square p-value:", test_result$p.value, 
        ifelse(test_result$p.value < 0.05, "(Significant)", "(Not significant)"), "\n")
  }
}

# Visualize relationships
plots_list <- list()
for (col in numerical_cols) {
  if (col %in% names(df)) {
    p <- ggplot(df, aes_string(x = "factor(target)", y = col, fill = "factor(target)")) +
      geom_violin(alpha = 0.7) +
      scale_fill_manual(values = c("skyblue", "coral"), labels = c("No Disease", "Disease")) +
      labs(title = paste(col, "by Target"),
           x = "Target",
           y = col,
           fill = "Target") +
      theme_minimal() +
      theme(legend.position = "none")
    plots_list[[col]] <- p
  }
}

# Save plots
png("../../results/bivariate_analysis.png", width = 1200, height = 800)
do.call(grid.arrange, c(plots_list, ncol = 2))
dev.off()

cat("\nBivariate analysis complete!\n")

File: Cardiovascular Disease Prediction using Machine Learning and Clinical Data/scripts/python/bivariate_analysis.py
"""
Bivariate Analysis

This script performs bivariate analysis on the Cardiovascular Disease Dataset.
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

# Set style
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (12, 6)

def load_data():
    """Load the dataset"""
    df = pd.read_csv('../../data/Cardiovascular_Disease_Dataset.csv')
    return df

def correlation_analysis(df):
    """Correlation analysis"""
    numerical_cols = ['age', 'restingBP', 'serumcholestrol', 'maxheartrate', 'oldpeak']
    correlation_matrix = df[numerical_cols].corr()
    
    print("=" * 80)
    print("CORRELATION ANALYSIS")
    print("=" * 80)
    print("\nCorrelation Matrix:")
    print(correlation_matrix)
    
    # Visualize
    plt.figure(figsize=(10, 8))
    sns.heatmap(correlation_matrix, annot=True, fmt='.3f', cmap='coolwarm', center=0)
    plt.title('Correlation Matrix - Numerical Variables')
    plt.tight_layout()
    plt.savefig('../../results/correlation_matrix.png')
    plt.close()

def numerical_vs_target(df):
    """Analyze numerical variables vs target"""
    numerical_cols = ['age', 'restingBP', 'serumcholestrol', 'maxheartrate', 'oldpeak']
    
    print("\n" + "=" * 80)
    print("NUMERICAL VARIABLES VS TARGET")
    print("=" * 80)
    
    for col in numerical_cols:
        if col in df.columns:
            print(f"\n{col.upper()} by Target:")
            grouped = df.groupby('target')[col].agg(['mean', 'median', 'std'])
            print(grouped)
            
            # Statistical test
            group_0 = df[df['target'] == 0][col].dropna()
            group_1 = df[df['target'] == 1][col].dropna()
            statistic, p_value = stats.ttest_ind(group_0, group_1)
            print(f"  t-test p-value: {p_value:.4f} ({'Significant' if p_value < 0.05 else 'Not significant'})")

def categorical_vs_target(df):
    """Analyze categorical variables vs target"""
    categorical_vars = ['gender', 'chestpain', 'fastingbloodsugar', 'restingrelectro', 
                        'exerciseangia', 'slope', 'noofmajorvessels']
    
    print("\n" + "=" * 80)
    print("CATEGORICAL VARIABLES VS TARGET")
    print("=" * 80)
    
    for var in categorical_vars:
        if var in df.columns:
            print(f"\n{var.upper()} vs TARGET:")
            crosstab = pd.crosstab(df[var], df['target'])
            print(crosstab)
            
            # Chi-square test
            chi2, p_value, dof, expected = stats.chi2_contingency(crosstab)
            print(f"  Chi-square p-value: {p_value:.4f} ({'Significant' if p_value < 0.05 else 'Not significant'})")

def visualize_relationships(df):
    """Visualize relationships"""
    numerical_cols = ['age', 'restingBP', 'serumcholestrol', 'maxheartrate', 'oldpeak']
    
    fig, axes = plt.subplots(2, 3, figsize=(18, 10))
    axes = axes.ravel()
    
    for idx, col in enumerate(numerical_cols):
        if col in df.columns and idx < len(axes):
            data_to_plot = [df[df['target'] == 0][col].dropna(), 
                            df[df['target'] == 1][col].dropna()]
            axes[idx].violinplot(data_to_plot, positions=[0, 1])
            axes[idx].set_xticks([0, 1])
            axes[idx].set_xticklabels(['No Disease', 'Disease'])
            axes[idx].set_title(f'{col} by Target')
            axes[idx].set_ylabel(col)
    
    if len(numerical_cols) < len(axes):
        fig.delaxes(axes[len(numerical_cols)])
    
    plt.tight_layout()
    plt.savefig('../../results/bivariate_analysis.png')
    plt.close()

def main():
    """Main function"""
    df = load_data()
    correlation_analysis(df)
    numerical_vs_target(df)
    categorical_vs_target(df)
    visualize_relationships(df)
    print("\nBivariate analysis complete!")

if __name__ == "__main__":
    main()

